### DISCUSSION TITLE
Query.filter() to accept filters as a list in clausal form

### DISCUSSION THREAD

**Discussion by fraser-langton** [sqlalchemy/sqlalchemy#7913] [Ideas]
### Describe the use case

Use case: **Pass list of nested filters in clausal form to Query.filer() to allow cleaner query construction**. 
  
Currently you may construct a filter by appending to a list, then apply to a query.filter() and they will all be taken as `and_`
eg
```python
query = session.query(models.Fruit)
query_filter = [
    models.Fruit.name.in_(('apple', 'banana', 'dragonfruit', 'tomato')),
    models.Fruit.colour == 'red',
    models.Fruit.is_yummy.is_(True)
]
query = query.filter(*query_filter)
```
will emit the sql:
```sql
SELECT fruit.tasks.id, fruit.tasks.name, fruit.tasks.colour, fruit.tasks.is_yummy 
FROM fruit.tasks 
WHERE fruit.tasks.name IN ('apple','banana','dragonfruit','tomato') AND fruit.tasks.colour = 'red' AND fruit.tasks.is_yummy IS true
```

Suggestion: extend query.filter() to accept list of lists (or single filter) so you may construct a series of and/or filters using a list
eg
```python
query = session.query(models.Fruit)
query_filter = [
    models.Fruit.name.in_(('apple', 'banana', 'dragonfruit', 'tomato')),
    [models.Fruit.colour == 'red', models.Fruit.is_yummy.is_(True)]
]
query = query.filter(*query_filter)
```
would emit the sql:
```sql
SELECT fruit.tasks.id, fruit.tasks.name, fruit.tasks.colour, fruit.tasks.is_yummy 
FROM fruit.tasks 
WHERE fruit.tasks.name IN ('apple','banana','dragonfruit','tomato') AND (fruit.tasks.colour = 'red' OR fruit.tasks.is_yummy IS true)
```

ie for list of conditions `[a, b, [c1, c2], [d1, d2, d3, d4]]` it is interpreted as `and_(a, b, or_(c1, c2), or_(d1, d2, d3, d4))`

### Databases / Backends / Drivers targeted

all

### Example Use

```python
query: Query = session.query(models.Task)
query_filter = []
if status is not None:
    query_filter.append(models.Task.status.in_(status))
if teams is not None:
    teams_filter = [models.Task.contractor_id.in_(teams)]  # this one becomes an or
    if unallocated_tasks:
        teams_filter.append([models.Task.contractor_id.is_(None)])
    query_filter.append(teams_filter)
query = query.filter(*query_filter)

print_query(query.filter(*query_filter))
```

**Comment by CaselIT** [comment]
Hi,

I'm personally -1 on this, since that api does not seem very clear. Also what would happen in this case `[a, b, [c, [d, e]]]`?

**Reply by fraser-langton** [reply]
It just occurred to me since `and` filters can be constructed entirely from lists, I'd argue the fact filter accepts a list is perhaps an unclear api, could it just as fairly construct the list into `or`? But I'd think there's a reason it is implemented as it is 🤷‍♂️

**Comment by zzzeek** [comment]
this use case is already suited by using `and_` / `or_` appropriately.     I wasn't even sure how to translate the example since without explicit use of and_/or_, i couldnt tell which parts should be AND and which should be OR.   

now if we want an existing and_() or or_() consturct to be more composable, like being able to append elements to it, we can do that, like `some_or.append(new_clause)`  or `some_or += new_clause`, that's more doable.

```py

from sqlalchemy import or_

query: Query = session.query(models.Task)
query_filter = []
if status is not None:
    query_filter.append(models.Task.status.in_(status))
if teams is not None:
    teams_filter = [models.Task.contractor_id.in_(teams)]
    if unallocated_tasks:
        teams_filter.append([models.Task.contractor_id.is_(None)])
    query_filter.append(or_(*teams_filter))
query = query.filter(*query_filter)

print_query(query.filter(*query_filter))
```

**Reply by fraser-langton** [reply]
> now if we want an existing and_() or or_() consturct to be more composable, like being able to append elements to it, we can do that, like `some_or.append(new_clause)` or `some_or += new_clause`, that's more doable.

👍 makes more sense than my suggestion, making the filter clauses more composable is what I was thinking about. 

```python
from sqlalchemy import or_

query: Query = session.query(models.Task)
query_filter = []
if status is not None:
    query_filter.append(models.Task.status.in_(status))
if teams is not None:
    teams_or = or_(models.Task.contractor_id.in_(teams))
    if unallocated_tasks:
        teams_or.append(models.Task.contractor_id.is_(None))
    query_filter.append(teams_or)
query = query.filter(*query_filter)

print_query(query.filter(*query_filter))
```

**Comment by enzymz** [comment]
```
db.MetaData.reflect(meta)
Session = sessionmaker(bind=engine)
session = Session()
obj = meta.tables[table_name]
filter_ = []
for k, v in kwargs.items():
    filter_.append((obj.columns.get(k) == v))
session.query(obj).filter(*filter_).all()
```
It's work for me
